### PR TITLE
Build precompiled and publish to S3

### PR DESCRIPTION
**Pull request by wingyplus**
Continuing #12038 from the discussion in https://github.com/elixir-lang/elixir/pull/12053#issuecomment-1208268123. I create a new workflow to publish a precompiled version to s3. Currently, it will run on every commit that pushes to main. If it's too much, we can change it to run every night (with a cron schedule).

### CODE REVIEW

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [inline .github/workflows/publish_precompiled.yml:45]: I copied the destination from https://github.com/hexpm/bob/blob/main/priv/scripts/elixir/elixir.sh#L81. But it can be change.

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [inline .github/workflows/publish_precompiled.yml:52]: What region should use?

**Comment by wingyplus** [comment]
These are files that this workflow publishes. The screenshot below is tested on my bucket.

Screen Shot 2565-09-06 at 21 57 58

**Review by wojtekmach — APPROVED** [review]
This is great! Besides the one comment I believe all that is left is s3 auth tokens and bumping fastly cache, both of which I can help making happen.

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml:45]: @wojtekmach I'll look into it. :)

I also have a question, do we still need to publish a shasum per file as the release workflow does?

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [inline .github/workflows/publish_precompiled.yml:38]: We should change the name to branch like `main-otp-${{ matrix.otp }}` to make it compatible with Bob right?

Comment (wingyplus) [inline .github/workflows/publish_precompiled.yml:38]: We should change the name to branch like main-otp-${{ matrix.otp }} to make it compatible with Bob right?

Comment (wojtekmach) [reply .github/workflows/publish_precompiled.yml:38]: Yeah, not sure. I think we can rename it to `main-otp-x` just before uploading to Bob.

**Comment by wingyplus** [comment]
@wojtekmach I addressed all issues that you commented. Please kindly review again. 🙇

**Comment by wingyplus** [comment]
The freebsd raise an error:

```
gmake compile
ld-elf.so.1: /lib/libc.so.7: version FBSD_1.7 required by /usr/local/lib/erlang24/erts-12.3.2.2/bin/beam.smp not found
gmake: *** [Makefile:83: lib/elixir/src/elixir_parser.erl] Error 1
```

I'm not sure why. :(

File: .github/workflows/publish_precompiled.yml
Comment (wojtekmach) [inline .github/workflows/publish_precompiled.yml:26]: 18.04 is deprecated:

```suggestion
    runs-on: ubuntu-20.04
```

File: .github/workflows/publish_precompiled.yml
Comment (wojtekmach) [inline .github/workflows/publish_precompiled.yml:50]: I'd remove ` || true` and `touch builds.txt` as it could obscure s3 network errors. It seems it is only useful for when there's no builds.txt in s3 already?

File: .github/workflows/publish_precompiled.yml
Comment (wojtekmach) [inline .github/workflows/publish_precompiled.yml:61]: I think you might have already asked about it, do we need to upload shasum files? They are redundant since we have builds.txt so maybe we don't?

**Review by wojtekmach — COMMENTED** [review]
cc @ericmj

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml:26]: Fixed

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml]: Yes it is

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml:60]: I’m not sure if we should keep it like release workflow does?

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml]: I changed to || touch builds.txt instead.:)

Comment (wojtekmach) [reply .github/workflows/publish_precompiled.yml]: If we got an intermittent network failure on `aws s3 cp` then we would create empty builds.txt whereas I think it is better to crash.

Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml]: Ahh, I see. I’ll take a look at it

File: .github/workflows/publish_precompiled.yml
Comment (ericmj) [inline .github/workflows/publish_precompiled.yml:55]:
```suggestion
          AWS_ACCESS_KEY_ID: ${{ secrets.HEXPM_ELIXIR_BUILDS_AWS_ACCESS_KEY_ID }}
          AWS_SECRET_ACCESS_KEY: ${{ secrets.HEXPM_ELIXIR_BUILDS_AWS_SECRET_ACCESS_KEY }}
```

**Review by ericmj — COMMENTED** [review]
This looks good, we just need to add the Fastly cache purging.

File: .github/workflows/publish_precompiled.yml
Comment (ericmj) [inline .github/workflows/publish_precompiled.yml:79]:
```suggestion
            -H "Fastly-Key: ${BOB_FASTLY_PURGE_KEY}" \
```

**Comment by wingyplus** [comment]
@wojtekmach @ericmj I addressed all comment suggestions. And I just added fastly purge cache after build but unfortunately, I've not fastly setup, so I can't test it by myself. :(

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [inline .github/workflows/publish_precompiled.yml:85]: I'm not sure what service name we should use or just use `secrets` instead?

**Comment by ericmj** [comment]
> I addressed all comment suggestions. And I just added fastly purge cache after build but unfortunately, I've not fastly setup, so I can't test it by myself. :(

Thank you. We can test this 👍

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml:79]: Fixed. 🙇

File: .github/workflows/publish_precompiled.yml
Comment (ericmj) [reply .github/workflows/publish_precompiled.yml:85]: Lets use secret here as well.

File: .github/workflows/publish_precompiled.yml
Comment (ericmj) [inline .github/workflows/publish_precompiled.yml:85]:
```suggestion
          BOB_FASTLY_KEY: ${{ secrets.HEXPM_REPO_FASTLY_KEY }}
          BOB_FASTLY_SERVICE: ${{ secrets.HEXPM_REPO_FASTLY_SERVICE }}
```

Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml]: @ericmj Changed. And rename BOB_FASTLY_* to just FASTLY_*. :)

**Comment by paulo-ferraz-oliveira** [comment]
Will this eventually mean we replace the place `setup-beam` fetches the pre-compiled from (https://github.com/erlef/setup-beam/blob/main/src/install-elixir.sh#L12)? Or is this just a normal redirect to S3 (thus no changes to `setup-beam`)?

**Comment by wingyplus** [comment]
> Will this eventually mean we replace the place `setup-beam` fetches the pre-compiled from (https://github.com/erlef/setup-beam/blob/main/src/install-elixir.sh#L12)? Or is this just a normal redirect to S3 (thus no changes to `setup-beam`)?

I’m expecting that there’s no changes on someone who fetch pre-compiled version. The file still put to the same place like before. :)

File: .github/workflows/publish_precompiled.yml
Comment (wojtekmach) [inline .github/workflows/publish_precompiled.yml:41]: repo.hex.pm also has main.zip (without -otp-X) which is compiled against oldest supported OTP. We should support it as well and then I think we can ship it.

**Comment by wingyplus** [comment]
Hi,

i just see the issue https://github.com/ScoopInstaller/Main/issues/3886#issuecomment-1239843242. I start thinking that we should bring back the shasum file per version for consumer usage reason. And it will consistent with current release workflow.

what do you think?

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml:41]: @wojtekmach So in this case it should build with otp24 right?

Comment (wojtekmach) [reply .github/workflows/publish_precompiled.yml:41]: Yes

**Comment by wojtekmach** [comment]
Per https://github.com/elixir-lang/elixir/issues/12134 scoop might go with repo.hex.pm/builds/elixir/* and perhaps theyd use builds.txt for checksums

File: .github/workflows/publish_precompiled.yml
Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml:41]: Let me try

Comment (wingyplus) [reply .github/workflows/publish_precompiled.yml:41]: @wojtekmach main.zip added in https://github.com/elixir-lang/elixir/pull/12118/commits/6a2c7d5dd75fdd3d93ee7734920c531dc313601f. Please kindly review.

**Comment by wingyplus** [comment]
> Looks great. We need to coordinate merging this with changing bob.

I've open a PR https://github.com/hexpm/bob/pull/125 to remove main. But it seems we need to have building doc feature before removing it. If have anything need to do more, feel free to comment. 🙇

**Comment by wojtekmach** [comment]
I believe docs is the remaining thing!

**Comment by wingyplus** [comment]
Yeah, i’ll try tackle that after this changes get merge.

**Comment by wojtekmach** [comment]
I believe we can only merge this after we stop Bob from building elixir main as not to keep overwriting builds.txt. But we can only stop bob from building main if we can build docs on GitHub. So I think we should do it as part of this PR. WDYT?

**Comment by wingyplus** [comment]
> I believe we can only merge this after we stop Bob from building elixir main as not to keep overwriting builds.txt. But we can only stop bob from building main if we can build docs on GitHub. So I think we should do it as part of this PR. WDYT?

I’m ok to continue on this PR.

**Pull request closed without merging**
